### PR TITLE
Application Management Publisher UI

### PR DESCRIPTION
**Pull request by menakaj**
This PR contains following changes.
1. Created auth handler component to get/ revoke and refresh tokens. 
2. Styling was done based on https://github.com/airbnb/javascript/tree/master/react#basic-rules
3. Apis are created for application and platform creation/ access and artifact upload/ download.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationKey.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * This holds api application consumer key and secret.
+ */
+@XmlRootElement
+public class ApiApplicationKey {
+    @XmlElement
+    private String client_id;
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationKey.java]: Can you change the attribute name with our usual camel case coding convensions?

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/AccessTokenInfo.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * This hold access token info that returned from the api call
+ */
+@XmlRootElement(name = "AccessTokenInfo")
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AccessTokenInfo {
+
+    @XmlElement(required = true, name = "token_type")
+    private String token_type;
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/AccessTokenInfo.java]: Shall we change it to camelCase?

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/service/impl/AuthHandlerServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                    .requestInterceptor(new BasicAuthRequestInterceptor(apiApplicationKey.getConsumerKey(),
+                            apiApplicationKey.getConsumerSecret()))
+                    .contract(new JAXRSContract()).encoder(new JacksonEncoder()).decoder(new JacksonDecoder())
+                    .target(TokenIssuerService.class, Constants.TOKEN_ENDPOINT);
+            AccessTokenInfo accessTokenInfo = tokenIssuerService.getToken(Constants.PASSWORD_GRANT_TYPE,
+                    userName, password, Constants.SCOPES);
+            JSONObject loginInfo = new JSONObject(accessTokenInfo);
+            loginInfo.append(Constants.USER_NAME, userName);
+            loginInfo.append(Constants.APPLICATION_INFO, new JSONObject(apiApplicationKey));
+            System.out.println(loginInfo);
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/service/impl/AuthHandlerServiceImpl.java:109]: Remove system.out.

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/OAuthRequestInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import feign.RequestInterceptor;
+import feign.RequestTemplate;
+
+import static feign.Util.checkNotNull;
+
+/**
+ * This is a request interceptor to add oauth token header.
+ */
+public class OAuthRequestInterceptor implements RequestInterceptor {
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/OAuthRequestInterceptor.java:30]: Can't we use the classes directly from the device mgt for these?

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiRegistrationProfile.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ */
+
+package org.wso2.carbon.device.application.mgt.auth.handler.util.dto;
+
+
+/**
+ * This class represents the data that are required to register
+ * the oauth application.
+ */
+public class ApiRegistrationProfile {
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiRegistrationProfile.java:26]: Can't we use the class from device-mgt core?

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationRegistrationService.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * This is the application registration service that exposed for apimApplicationRegistration
+ */
+
+@Path("/register")
+public interface ApiApplicationRegistrationService {
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationRegistrationService.java:32]: Do we need to have separate API for this??

File: components/application-mgt/org.wso2.carbon.device.application.mgt.publisher.ui/src/main/resources/publisher/src/api/applicationMgtApi.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+     * Api for create an application.
+     * @param: applicationData: The application data object. This contains an object array of each step data from
+     * application creation wizard.
+     *
+     * From applicationData, the proper application object will be created and send it to the api.
+     * */
+    static createApplication(applicationData) {
+        let {application, images} = Helper.buildApplication(applicationData);
+        const headers = AuthHandler.createAuthenticationHeaders("application/json");
+        console.log(application);
```
Comment (sinthuja) [inline components/application-mgt/org.wso2.carbon.device.application.mgt.publisher.ui/src/main/resources/publisher/src/api/applicationMgtApi.js:41]: Remove console.log

**Review by sinthuja — CHANGES_REQUESTED** [review]
In authentication handler component, I think we can use some classes directly from the device-mgt core, without having to duplicate the code. And there are some naming convensions issues. Please do fix them.

File: components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationRegistrationService.java (hunk 6 above)
Comment (menakaj) [reply components/application-mgt/org.wso2.carbon.device.application.mgt.authhandler/src/main/java/org/wso2/carbon/device/application/mgt/auth/handler/util/dto/ApiApplicationRegistrationService.java:32]: This is an interface that should be passed to the Feign client. We do not implement this class. In Feign, we provide the base url, the service interface and call the interface method.
Ex: To get access token, we provide the interface TokenIssuerService.java which has the path "/token" and two methods, getToken and getRefreshToken. 
In Feign it calls BASE_URL/token, with the parameters provided to the method.

**Review by sinthuja — APPROVED** [review]

**Pull request merged**